File: sdk/src/decode.ts
import type { AccountTypeDef, FieldDef } from "./generated/accounts";
import { ACCOUNT_TYPES_BY_DISC } from "./generated/accounts";
import { encodePk } from "./bytes";

export type DecodedField = {
  name: string;
  type: string;
  offset: number;
  confidence: "certain" | "inferred";
  value: unknown;
};

function parseFixedBytes(t: string): number | null {
  const m = t.match(/^\[u8;(\d+)\]$/);
  return m ? Number(m[1]) : null;
}

export function decodeField(buf: Buffer, f: FieldDef): unknown {
  const o = f.offset;
  if (o >= buf.length) return undefined;
  const t = f.type;
  if (t === "Pubkey") {
    if (o + 32 > buf.length) return undefined;
    return encodePk(buf.subarray(o, o + 32));
  }
  if (t === "u8" || t === "u8/bool" || t === "bool") return buf[o];
  if (t === "u16") {
    if (o + 2 > buf.length) return undefined;
    return buf.readUInt16LE(o);
  }
  if (t === "u32") {
    if (o + 4 > buf.length) return undefined;
    return buf.readUInt32LE(o);
  }
  if (t === "u64") {
    if (o + 8 > buf.length) return undefined;
    return buf.readBigUInt64LE(o);
  }
  if (t === "i64") {
    if (o + 8 > buf.length) return undefined;
    return buf.readBigInt64LE(o);
  }
  const n = parseFixedBytes(t);
  if (n !== null) {
    if (o + n > buf.length) return buf.subarray(o);
    return buf.subarray(o, o + n);
  }
  if (t === "RemainingBytes") return buf.subarray(o);
  return buf.subarray(o);
}

export function identifyAccount(data: Buffer): AccountTypeDef | undefined {
  if (data.length < 8) return undefined;
  return ACCOUNT_TYPES_BY_DISC[data.subarray(0, 8).toString("hex")];
}

export function decodeAccount(data: Buffer, type?: AccountTypeDef) {
  const t = type ?? identifyAccount(data);
  const disc = data.subarray(0, 8).toString("hex");
  if (!t) return { type: "unknown", disc, fields: [] as DecodedField[] };
  return {
    type: t.name,
    disc,
    fields: t.fields.map((f) => ({
      name: f.name,
      type: f.type,
      offset: f.offset,
      confidence: f.confidence,
      value: decodeField(data, f),
    })),
  };
}

export function fieldMap(decoded: { fields: DecodedField[] }): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const f of decoded.fields) out[f.name] = f.value;
  return out;
}

File: sdk-rs/src/decode.rs
use crate::bytes::encode_pk;
use crate::catalog::{field_offset, AccountTypeDef, FieldDef};

#[derive(Debug, Clone)]
pub struct DecodedField {
    pub name: String,
    pub ty: String,
    pub offset: usize,
    pub confidence: String,
    pub value: DecodedValue,
}

#[derive(Debug, Clone)]
pub enum DecodedValue {
    Pubkey(String),
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    I64(i64),
    Bytes(Vec<u8>),
    Missing,
}

pub fn identify_account<'a>(data: &[u8], types: &'a [AccountTypeDef]) -> Option<&'a AccountTypeDef> {
    if data.len() < 8 {
        return None;
    }
    let disc = hex::encode(&data[..8]);
    types.iter().find(|t| t.discriminator == disc)
}

mod hex {
    pub fn encode(b: &[u8]) -> String {
        b.iter().map(|x| format!("{x:02x}")).collect()
    }
}

pub fn decode_field(data: &[u8], f: &FieldDef) -> DecodedValue {
    let Some(o) = field_offset(f) else {
        return DecodedValue::Missing;
    };
    if o >= data.len() {
        return DecodedValue::Missing;
    }
    let t = f.ty.as_str();
    if t == "Pubkey" {
        if o + 32 > data.len() {
            return DecodedValue::Missing;
        }
        return DecodedValue::Pubkey(encode_pk(&data[o..o + 32]));
    }
    if t == "u8" || t == "u8/bool" || t == "bool" {
        return DecodedValue::U8(data[o]);
    }
    if t == "u16" && o + 2 <= data.len() {
        return DecodedValue::U16(u16::from_le_bytes(data[o..o + 2].try_into().unwrap()));
    }
    if t == "u32" && o + 4 <= data.len() {
        return DecodedValue::U32(u32::from_le_bytes(data[o..o + 4].try_into().unwrap()));
    }
    if t == "u64" && o + 8 <= data.len() {
        return DecodedValue::U64(u64::from_le_bytes(data[o..o + 8].try_into().unwrap()));
    }
    if t == "i64" && o + 8 <= data.len() {
        return DecodedValue::I64(i64::from_le_bytes(data[o..o + 8].try_into().unwrap()));
    }
    if let Some(n) = t.strip_prefix("[u8;").and_then(|s| s.strip_suffix(']')).and_then(|s| s.parse::<usize>().ok()) {
        let end = (o + n).min(data.len());
        return DecodedValue::Bytes(data[o..end].to_vec());
    }
    if t == "RemainingBytes" {
        return DecodedValue::Bytes(data[o..].to_vec());
    }
    DecodedValue::Bytes(data[o..].to_vec())
}

pub fn decode_account(data: &[u8], ty: &AccountTypeDef) -> Vec<DecodedField> {
    ty.fields
        .iter()
        .filter_map(|f| {
            let o = field_offset(f)?;
            Some(DecodedField {
                name: f.name.clone(),
                ty: f.ty.clone(),
                offset: o,
                confidence: f.confidence.clone(),
                value: decode_field(data, f),
            })
        })
        .collect()
}
